src/units/abilities.cpp

```diff
@@ -1168,52 +1168,137 @@ void attack_type::modified_attacks(unsigned & min_attacks,
 	}
 }
 
-//Functions used for change damage_type list with damage
-static std::optional<std::string> select_damage_type(const unit_ability_list& abil_list, const std::string& type)
+//functions for select damage type who must be returned for each attribute used in damage_type()
+static std::string select_replacement_type(const unit_ability_list& damage_type_list)
 {
+	std::unordered_map<std::string, unsigned int> type_count;
+	unsigned int max = 0;
+	for(auto& i : damage_type_list) {
+		const config& c = *i.ability_cfg;
+		if(c.has_attribute("replacement_type")) {
+			std::string type = c["replacement_type"].str();
+			unsigned int count = ++type_count[type];
+			if((count > max)) {
+				max = count;
+			}
+		}
+	}
+
+	if (type_count.empty()) return "";
+
 	std::vector<std::string> type_list;
-	for(auto& i : abil_list) {
-		if(!(*i.ability_cfg)[type].str().empty()){
-			type_list.push_back((*i.ability_cfg)[type].str());
+	for(auto& i : type_count){
+		if(i.second == max){
+			type_list.push_back(i.first);
 		}
 	}
-	if(type_list.size() >= 2){
+
+	if(type_list.size() > 1){
 		std::sort(type_list.begin(), type_list.end());
-		if(type_list.size() >= 3){
-			std::unordered_map<std::string, unsigned int> type_count;
-			for( const std::string& character : type_list ){
-				type_count[character]++;
+	}
+
+	if(!type_list.empty()){
+		return type_list.front();
+	}
+	return "";
+}
+
+static std::string select_alternative_type(const unit_ability_list& damage_type_list, unit_ability_list resistance_list, const unit& u)
+{
+	std::map<std::string, int> type_res;
+	//for don't have bug in showing in pre_attack window if opponent has resistance greater what 100% type_res shouldn't be initialised.
+	std::optional<int> max_res;
+	for(auto& i : damage_type_list) {
+		const config& c = *i.ability_cfg;
+		if(c.has_attribute("alternative_type")) {
+			std::string type = c["alternative_type"].str();
+			if(type_res.count(type) == 0){
+				type_res[type] = u.resistance_value(resistance_list, type);
+				max_res = max_res ? std::max(*max_res, type_res[type]) : type_res[type];
+			}
+		}
+	}
+
+	if (type_res.empty()) return "";
+
+	std::vector<std::string> type_list;
+	if(max_res){
```

Comment:
`max_res` cannot be nullopt here.

Similar simplification for the rest of the function as above.